edgexfoundry/edgex-go
# fix(snap): make secretstore connect hook idempotent

Fixes #3814 by copying the tokens instead of moving them away from the origin.

In addition, it improves logging:
- The list of directories are now fully printed (not just the first line)
- Failed connect operation returns an error (when token not present because edgex-go doesn't provide it or someone has remove it)

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>







**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:


## Testing Instructions
See #3814 to check how it used to work.

Using a locally-build edgexfoundry, so the plug doesn't connect automatically.
```
# install device-camera
$ sudo snap install edgex-device-camera --edge
edgex-device-camera (edge) 2.0.2-dev.4 from Canonical✓ installed
# connect the plug
$ sudo snap connect edgexfoundry:edgex-secretstore-token edgex-device-camera:edgex-secretstore-token
# check that token is copied
$ sudo ls /var/snap/edgex-device-camera/current/device-camera/secrets-token.json
/var/snap/edgex-device-camera/current/device-camera/secrets-token.json
# remove device-camera
$ sudo snap remove edgex-device-camera
edgex-device-camera removed

# install device-camera again
$ sudo snap install edgex-device-camera --edge
edgex-device-camera (edge) 2.0.2-dev.4 from Canonical✓ installed
# connect the plug again
$ sudo snap connect edgexfoundry:edgex-secretstore-token edgex-device-camera:edgex-secretstore-token
# check that token is copied
$ sudo ls /var/snap/edgex-device-camera/current/device-camera/secrets-token.json
/var/snap/edgex-device-camera/current/device-camera/secrets-token.json
```

Test error handling:
```
# disconnect the plug
$ sudo snap disconnect edgexfoundry:edgex-secretstore-token edgex-device-camera:edgex-secretstore-token
# remove the token manually
$ sudo rm -v /var/snap/edgexfoundry/current/secrets/device-camera/secrets-token.json
removed '/var/snap/edgexfoundry/current/secrets/device-camera/secrets-token.json'
# connect the plug
$ sudo snap connect edgexfoundry:edgex-secretstore-token edgex-device-camera:edgex-secretstore-token
error: cannot perform the following tasks:
- Run hook prepare-plug-edgex-secretstore-token of snap "edgexfoundry" (run hook "prepare-plug-edgex-secretstore-token": <13>Nov 12 19:54:21 root: edgex-secretstore-token: could not find token for device-camera)
```

Look at the journals:
```
$ journalctl -n 100 -o cat | grep edgex-secretstore-token:
edgex-secretstore-token: prepare SLOTDIRS=[ "$SNAP_DATA/device-camera" ]
edgex-secretstore-token: prepare device-camera
edgex-secretstore-token: copying /var/snap/edgexfoundry/x1/secrets/device-camera/secrets-token.json to /tmp/device-camera-secrets-token.json
edgex-secretstore-token: connect SLOTDIRS=[ "$SNAP_DATA/device-camera" ]
edgex-secretstore-token: connect device-camera
edgex-secretstore-token: moving /tmp/device-camera-secrets-token.json to /var/snap/edgexfoundry/x1/secrets/device-camera/secrets-token.json
...
edgex-secretstore-token: prepare SLOTDIRS=[ "$SNAP_DATA/device-camera" ]
edgex-secretstore-token: prepare device-camera
edgex-secretstore-token: could not find token for device-camera
```
## New Dependency Instructions (If applicable)
